src/ints/mouse.cpp

```diff
@@ -18,1549 +18,932 @@
-	mouse.clipy = (int16_t)((Bits)CurMode->sheight - 1);
-
-	/* might be vidmode == 0x13?2:1 */
-	int16_t xratio = 640;
-	if (CurMode->swidth > 0)
-		xratio /= CurMode->swidth;
-	if (xratio == 0)
-		xratio = 1;
-
-	RestoreCursorBackground();
-
-	SaveVgaRegisters();
-
-	// Save Background
-	int16_t x, y;
-	uint16_t addx1, addx2, addy;
-	uint16_t dataPos = 0;
-	int16_t x1       = POS_X / xratio - mouse.hotx;
-	int16_t y1       = POS_Y - mouse.hoty;
-	int16_t x2       = x1 + CURSORX - 1;
-	int16_t y2       = y1 + CURSORY - 1;
-
-	ClipCursorArea(x1, x2, y1, y2, addx1, addx2, addy);
-
-	dataPos = addy * CURSORX;
-	for (y = y1; y <= y2; y++) {
-		dataPos += addx1;
-		for (x = x1; x <= x2; x++) {
-			INT10_GetPixel(x, y, mouse.page, &mouse.backData[dataPos++]);
-		};
-		dataPos += addx2;
-	};
-	mouse.background = true;
-	mouse.backposx   = POS_X / xratio - mouse.hotx;
-	mouse.backposy   = POS_Y - mouse.hoty;
-
-	// Draw Mousecursor
-	dataPos = addy * CURSORX;
-	for (y = y1; y <= y2; y++) {
-		uint16_t scMask = mouse.screenMask[addy + y - y1];
-		uint16_t cuMask = mouse.cursorMask[addy + y - y1];
-		if (addx1 > 0) {
-			scMask <<= addx1;
-			cuMask <<= addx1;
-			dataPos += addx1;
-		};
-		for (x = x1; x <= x2; x++) {
-			uint8_t pixel = 0;
-			// ScreenMask
-			if (scMask & HIGHESTBIT)
-				pixel = mouse.backData[dataPos];
-			scMask <<= 1;
-			// CursorMask
-			if (cuMask & HIGHESTBIT)
-				pixel = pixel ^ 0x0F;
-			cuMask <<= 1;
-			// Set Pixel
-			INT10_PutPixel(x, y, mouse.page, pixel);
-			dataPos++;
-		};
-		dataPos += addx2;
-	};
-	RestoreVgaRegisters();
-}
+} MouseEvent;
 
-static void CursorMoved(float xrel, float yrel, float x, float y, bool emulate)
+static class MouseQueue final {
+public:
+    MouseQueue();
+
+    void AddEvent(MouseEvent &event);
+    void FetchEvent(MouseEvent &event);
+    void ClearEventsDOS();
+    void StartTimerIfNeeded();
+    void Tick();
+
+    uint8_t start_delay_ps2     = 5; // for PS/2 events, in milliseconds
+    uint8_t start_delay_dos_btn = 1; // for DOS button events
+    uint8_t start_delay_dos_mov = 5; // for DOS move/wheel events
+
+private:
+    MouseQueue(const MouseQueue &)            = delete;
+    MouseQueue &operator=(const MouseQueue &) = delete;
+
+    void AggregateEventsDOS(MouseEvent &event);
+    MouseEvent &PopEventBtn();
+    void UpdateDelayCounters();
+
+    std::array<MouseEvent, 8> events = {}; // a modulo queue of events
+    uint8_t idx_first                = 0;  // index of the first event
+    uint8_t num_events               = 0;  // number of events in the queue
+
+    MouseButtons12S last_buttons_12S = 0; // last buttons reported
+
+    bool queue_overflow    = false;
+    bool timer_in_progress = false;
+
+    // Time in milliseconds which has to elapse before event can take place
+    uint8_t delay_ps2     = 0;
+    uint8_t delay_dos_btn = 0; // for DOS button events
+    uint8_t delay_dos_mov = 0; // for DOS move/wheel events
```

Comment:
button, move?

Reply:
I'd suggest `_ms` suffixes for all these.

Reply:
There's quite a few variable-triplets here. Would it be better to logically group them in structs? Then you could refer to them as `delay.dos_button_ms`, for example, intialise the structs in one go, etc.

Reply:
...or maybe you could slice it in a different way: `ps2`, `dos_moved` and `dos_wheel` could be three instances of the same struct that contains `event`, `delay_ms`, `start_delay_ms`, etc. I think that would be my preference as it would be more logical and you'd get rid of the copy-paste style repetition of names.

Reply:
These structures are now "symmetric" - there are two separate `event` variables for a single `delay` variable (mouse move + wheel), and for the button events we have a queue.

But added `_ms` suffix for delays,. and put them into structs.